router/acme/acme.go

```diff
@@ -0,0 +1,480 @@
+		return
+	}
+
+	// satisfy the http-01 challenge
+	if err := s.satisfyHTTP01Challenge(cert, challenge); err != nil {
+		s.setFailed(cert, "error satisfying http-01 challenge: %s", err)
+		return
+	}
+
+	// finalize the order
+	order, keyDER, err := s.finalizeOrder(cert, order)
+	if err != nil {
+		s.setFailed(cert, "error finalizing order: %s", err)
+		return
+	}
+
+	// set the current certificate to the newly issued certificate
+	if err := s.setCertificate(cert, keyDER, order); err != nil {
+		s.setFailed(cert, "error setting current certificate: %s", err)
+		return
+	}
+
+	return
+}
+
+func (s *Service) getOrCreateOrder(cert *ct.ManagedCertificate) (*acme.Order, error) {
+	log := s.log.New("domain", cert.Domain)
+
+	// if the managed certificate already has an OrderURL, fetch it and
+	// check if it's valid
+	if url := cert.OrderURL; url != "" {
+		log.Info("checking existing order", "order.url", url)
+		order, err := s.client.FetchOrder(s.account, url)
+		if err != nil {
+			log.Error("error checking existing order", "order.url", url, "err", err)
+		} else if order.Status == "invalid" {
+			log.Error("existing order is invalid", "order.url", url)
+		} else {
+			return &order, nil
+		}
+	}
+
+	// the managed certificate either has no OrderURL, or the order the
+	// OrderURL points to is not valid, so create a new order
+	log.Info("creating new order")
+	ids := []acme.Identifier{{Type: "dns", Value: cert.Domain}}
+	order, err := s.client.NewOrder(s.account, ids)
+	if err != nil {
+		log.Error("error creating new order", "err", err)
+		return nil, err
+	}
+
+	log.Info("updating managed certificate with new order", "order.url", order.URL)
+	cert.OrderURL = order.URL
+	if err := s.controller.UpdateManagedCertificate(cert); err != nil {
+		log.Error("error updating managed certificate with new order", "order.url", order.URL, "err", err)
+		return nil, err
+	}
+
+	return &order, nil
+}
+
+func (s *Service) getHTTP01Challenge(cert *ct.ManagedCertificate, order *acme.Order) *acme.Challenge {
+	log := s.log.New("domain", cert.Domain)
+
+	log.Info("getting http-01 challenge")
+	for _, authURL := range order.Authorizations {
+		auth, err := s.client.FetchAuthorization(s.account, authURL)
+		if err != nil {
+			continue
+		}
+		if challenge, ok := auth.ChallengeMap[acme.ChallengeTypeHTTP01]; ok {
+			log.Info("using http-01 challenge", "challenge.token", challenge.Token, "challenge.url", challenge.URL)
+			return &challenge
+		}
+	}
+	log.Error("missing http-01 challenge")
+	return nil
+}
+
+func (s *Service) satisfyHTTP01Challenge(cert *ct.ManagedCertificate, challenge *acme.Challenge) error {
+	log := s.log.New("domain", cert.Domain, "challenge.token", challenge.Token)
+
+	log.Info("configuring http-01 challenge response")
+	return s.responder.RespondHTTP01(cert, challenge, func() error {
+		// make multiple attempts to complete the challenge
+		attempts := attempt.Strategy{
+			Total: 60 * time.Second,
+			Delay: 1 * time.Second,
+		}
```

Comment:
We should only retry like this for 5xx.

Validations should retry every 13 minutes, as there is a limit of 5 failed validations per domain per hour and we need to handle the case where DNS is set up later, potentially several hours later.

For both issuance, renewals, and validation retries it may make sense to have a "task" data structure each with a scheduled run timestamp. There are two ways to model this, one involves locking in the database (we could use que-go for this) so we can have multiple `acme` workers with no leader. Alternatively we can do a scheduler-style leader lock, but we have to be even more careful about tasks being idempotent because we can't get a guarantee that there is only one leader due to the way consensus works.

Reply:
Worth noting that OCSP stapling will be a task on a schedule too, so I think we do need an abstraction to handle it.